docs/bpu_design_document/BPUDesignDoc.adoc

```diff
@@ -0,0 +1,636 @@
+| XiangShan Branch Predictor| https://github.com/OpenXiangShan/XiangShan-doc/blob/main/docs/frontend/bp.md
+|===
+
+[[OVERVIEW]]
+== OVERVIEW
+
+Branch Prediction Unit is the unit responsible for predicting the branch
+outcomes and branch targets before branch instructions are actually
+resolved in the pipeline in order to reduce latency between a branch and a
+subsequent instruction.
+
+It receives prediction input in the form of PC and instruction type from the
+Fetch unit and predicts the direction and target address from which Fetch unit
+has to get the next instruction.
+
+
+[[Overview_Block_Diagram]]
+=== Overview Block Diagram
+
+image:media/bpu_overview.png[image,width=576,height=366]
+
+Figure 1 - Overview of BPU in reference to Fetch unit
+
+[[Functional_Description]]
+== Functional Description
+
+Branch Prediction Unit in Olympia is a two-tiered branch predictor where the
+first tier provides a simple but fast prediction. The second tier consists
+of a more accurate predictor which can predict even complex branches but takes an
+additional cycle.
+
+[[Input_Ports_BPU]]
+=== Input Ports in BPU
+
+**  `in_fetch_prediction_credits` - To receive credits from Fetch unit to indicate availability
+    of slots
+**  `in_fetch_prediction_req` - To receive prediction requests from Fetch unit
+
+**  `TODO-`: input port to receive de-allocated FTQ entry once that instruction packet is committed.
+
+[[Corresponding_output_ports]]
+==== Corresponding Output Ports
+
+** `out_bpu_prediction_credits` - in Fetch.cpp. To send credits to BPU to indicate availability of
+   slots.
+
+** `out_bpu_prediction_req` - in Fetch.cpp. To send PredictionInput to BPU.
+
+
+[[Output_Ports_BPU]]
+=== Output Ports in BPU
+
+** `out_fetch_prediction_res` - To send prediction result to Fetch unit
+
+[[Corresponding_input_ports]]
+==== Corresponding Input Ports
+
+** `in_bpu_prediction_res` - in Fetch.cpp. To receive prediction result from BPU.
+
+[[Counters]]
+=== Counters
+
+** `pred_req_num` - Total number of prediction requests made to BPU
+** `num_mispred` - Total number of mis-predictions
+** `mispred_perc` - Percentage of mis-predictions
+
+[[Unit_Block_Diagram]]
+=== Unit Block Diagram
+
+image:media/bpu_uarch.png[image,width=800,height=366]
+
+Figure 2 - Unit block diagram of BPU
+
+[[Block_Diagram_Description]]
+=== Block Diagram Description
+
+BPU consists of a BasePredictor (which is accessed in the
+first fetch cycle), a TAGE_SC_L-Predictor (which is also accessed
+in the first fetch cycle), and an FTQ (a simple
+queue) which stores the information of all in-flight branch instructions
+and is required in order to update the BPU once instruction are finally
+committed.
+
+[[Description_of_PredictionInput]]
+== Description of PredictionInput
+
+Olympia's Branch Predictor API intends the implementation of Branch Predictor to
+define custom PredictionInput
+
+[[Overview_of_PredictionInput]]
+===  Overview
```

Comment:
Are there header formatting issues with the document? The organization from this point on does not follow a progression I understand.